storage.go

```diff
@@ -0,0 +1,355 @@
+	err := s.Put(ctx, key, value)
+	if err != nil {
+		microerror.Mask(err)
+	}
+	return nil
+}
+
+func (s *Storage) Put(ctx context.Context, key, value string) error {
+	var body []byte
+	{
+		v := struct {
+			Data map[string]string `json:"data"`
+		}{
+			Data: map[string]string{
+				key: value,
+			},
+		}
+
+		var err error
+		body, err = json.Marshal(&v)
+		if err != nil {
+			return microerror.Maskf(err, "marshaling %#v", v)
+		}
+	}
+
+	_, err := s.k8sClient.Core().RESTClient().
+		Patch(types.MergePatchType).
+		Context(ctx).
+		AbsPath(s.tpoEndpoint).
+		Body(body).
+		DoRaw()
+	if err != nil {
+		return microerror.Maskf(err, "creating value for key=%s, patch=%s", key, body)
+	}
+
+	return nil
+}
+
+func (s *Storage) Exists(ctx context.Context, key string) (bool, error) {
+	data, err := s.getData(ctx)
+	if err != nil {
+		return false, microerror.Maskf(err, "Exists")
+	}
+
+	_, ok := data[key]
+	return ok, nil
+}
+
+func (s *Storage) Search(ctx context.Context, key string) (string, error) {
+	data, err := s.getData(ctx)
+	if err != nil {
+		return "", microerror.Maskf(err, "Search")
+	}
+
+	v, ok := data[key]
+	if !ok {
+		return "", microerror.Maskf(microstorage.NotFoundError, "Search key=%", key)
+	}
+
+	return v, nil
+}
+
+func (s *Storage) List(ctx context.Context, key string) ([]string, error) {
+	data, err := s.getData(ctx)
+	if err != nil {
+		return nil, microerror.Maskf(err, "Exists")
+	}
+
+	var list []string
+
+	keyLen := len(key)
+	for k, _ := range data {
+		if !strings.HasPrefix(k, key) {
+			continue
+		}
+
+		// k must be exact match or be separated with /.
+		// I.e. /foo is under /foo/bar but not under /foobar.
+		if len(k) != keyLen && k[keyLen] != '/' {
+			continue
+		}
+
+		list = append(list, k[keyLen+1:])
+	}
+
+	return list, nil
+}
+
+func (s *Storage) Delete(ctx context.Context, key string) error {
+	var body []byte
+	{
+		v := struct {
+			Data map[string]*string `json:"data"`
+		}{
+			Data: map[string]*string{
+				key: nil,
+			},
+		}
+
+		var err error
+		body, err = json.Marshal(&v)
+		if err != nil {
+			return microerror.Maskf(err, "marshaling %#v", v)
+		}
+	}
+
+	_, err := s.k8sClient.Core().RESTClient().
+		Patch(types.MergePatchType).
+		Context(ctx).
+		AbsPath(s.tpoEndpoint).
+		Body(body).
+		DoRaw()
+	if err != nil {
+		return microerror.Maskf(err, "deleting value for key=%s, patch=%s", key, body)
+	}
+
+	return nil
+}
+
+func (s *Storage) getData(ctx context.Context) (map[string]string, error) {
+	res, err := s.k8sClient.Core().RESTClient().
+		Get().
+		Context(ctx).
+		AbsPath(s.tpoEndpoint).
+		DoRaw()
+	if err != nil {
+		return nil, microerror.Maskf(err, "get TPO")
+	}
+
+	var v customObject
+	err = json.Unmarshal(res, &v)
+	if err != nil {
+		return nil, microerror.Maskf(err, "unmarshal TPO")
+	}
+
+	return v.Data, nil
+}
+
+func (s *Storage) log(v ...interface{}) {
+	a := make([]interface{}, 0, len(v)+len(s.logctx))
+	a = append(a, v...)
+	a = append(a, s.logctx...)
+	s.logger.Log(a...)
```

Comment:
I can imagine that this breaks the log location. I would rather like to implement a context based logger that maybe wraps the logger, aligns the log location, and can be used as usual without the need of this private `log` function here.